Define stream limits as counts (#1906)
quicwg/base-drafts: draft-ietf-quic-transport.md

QUIC allows for an arbitrary number of streams to operate concurrently.  An
endpoint limits the number of concurrently active incoming streams by limiting

> the number of streams (see {{stream-limit-increment}}).

Reviewer: ...of each type.

The stream limit is specific to each endpoint and applies only to the peer that
receives the setting. That is, the client limits the number of streams the

> server can initiate, and the server limits the number of streams the client can

Reviewer: Probably want to mention "of each type" in this paragraph as well.

> receives the setting. That is, clients limit the number of streams the server

Reviewer 2: Mixing plural clients with singular server. And vice versa below.

> receives the setting. That is, clients limit the number of streams the server

Reviewer 3: Hm I'm not in love with the "number of streams" terminology. With "maximum stream ID" it was very obvious what was being limited. If the max stream ID was X, you clearly could not open a steam with ID X+1 (or X+4 but whatever :>).

With a limit on the "number of a streams" this is less clear. If a server limits the client to 10, presumably the idea is that this is equivalent to a max stream id limit of the 10th client initiated stream ID. But it also seems like you could read this limit to permit the client to open any 10 steams. Like 0, 40, 80, 120, ... I think this is not the intent, but the terminology makes this much less clear to me.

For example, the text below:

"A peer MUST NOT initiate a stream with a higher stream ID than the largest limit
maximum stream ID it has received. "

Clearly the limit applies to stream ID, but using "number of streams" seems to obscure that.

Author: I agree that this isn't great, but it's an artifact of encoding stream type in the ID.  A decision I really don't want to revisit, BTW.

Given that the number of streams can be very easily determined from the ID with `(id+3)/4`, or the test for whether an stream can be opened is `id/4 >= limit`, I think that this is quite workable.

I've tweaked the text you point at.  Hopefully this is clearer.

initiate.  Each endpoint may respond on streams initiated by the other peer,
regardless of whether it is permitted to initiate new streams.

Endpoints MUST NOT exceed the limit set by their peer.  An endpoint that
receives a STREAM frame with an ID greater than the limit it has sent MUST treat
this as a stream error of type STREAM_LIMIT_ERROR ({{error-handling}}), unless

> this is a result of a change in the initial limits (see {{zerortt-parameters}}).

Reviewer: Existing text, but....  We don't allow reductions in TPs after accepting 0-RTT, so how would it be the result of a change in the initial limits?

Author: Cutting that.  I didn't think much on it before, but you are right that it's nonsensical.

## Stream Limit Increment {#stream-limit-increment}

An endpoint limits the number of concurrently active incoming streams.  An
initial value is set in the transport parameters (see
{{transport-parameter-definitions}}) and is subsequently increased by
MAX_STREAMS frames (see {{frame-max-streams}}).  Separate limits apply to
bidirectional and unidirectional streams.

As with stream and connection flow control, this document leaves when and how
many streams to make available to a peer via MAX_STREAMS to implementations.

> 

Reviewer 4: Are we fine with removing this sentence?

I ask this because it is the only reliable way the things would work. Adding a sentence below about an optimization (i.e. using STREAMS_BLOCKED as a signal to send to MAX_STREAMS; it is merl) at the same time removing the required interaction might confuse the readers.

Author: Yeah, I think that I might restore this, but attach it to the paragraph above.

Reviewer: And since I see the text above, you've presumably done this.

The STREAMS_BLOCKED frame ({{frame-streams-blocked}}) can be used to signal a

> shortage of available streams. Implementations could use this as a signal to

Reviewer 5: I think shortage is not quite specific enough.  To me, that might mean "I only have 5 available" which I don't believe is the intent?  How about "The STREAMS_BLOCKED frame ({{frame-streams-blocked}}) signals that a new stream could not be created.  Implementations can use this as a signal to the peer to send a MAX_STREAMS frame with a larger limit."

Author: Thanks.  I struggled with that.

## Stream Limit Increment {#stream-limit-increment}

An endpoint limits the number of concurrently active incoming streams.  An
initial value is set in the transport parameters (see
{{transport-parameter-definitions}}) and is subsequently increased by
MAX_STREAMS frames (see {{frame-max-streams}}).  Separate limits apply to
bidirectional and unidirectional streams.

As with stream and connection flow control, this document leaves when and how
many streams to make available to a peer via MAX_STREAMS to implementations.
Implementations might choose to increase limits as streams close to keep the
number of streams available to peers roughly consistent.

> The STREAMS_BLOCKED frame ({{frame-streams-blocked}}) signals that a new stream

Reviewer: Ugh.  So we now have STREAM_BLOCKED and STREAMS_BLOCKED?  No one is *ever* going to get those backward, no....

Author: I had exactly the same thought.  It's just that I can't think of a better name.

Reviewer: Perhaps, in the interest of consistency....
- MAX_DATA => DATA_BLOCKED
- MAX_STREAMS => STREAMS_BLOCKED
- MAX_STREAM_DATA => STREAM_DATA_BLOCKED

Author: \o/ thanks.

  endpoint SHOULD stop sending MAX_STREAM_DATA frames when the receive stream
  enters a "Size Known" state.

> * The limit on streams for a stream of a given type is sent in MAX_STREAMS

Reviewer 5: ```suggestion
* The limit on streams of a given type is sent in MAX_STREAMS
```

a change in the initial limits (see {{zerortt-parameters}}).

## MAX_STREAMS Frame {#frame-max-streams}

The MAX_STREAMS frame (type=0x1c and 0x1d) informs the peer of the number of
streams it is permitted to open.  A MAX_STREAMS frame with a type of 0x1c
applies to bidirectional streams; a MAX_STREAMS frame with a type of 0x1d
applies to unidirectional streams.

The frame is as follows:

~~~
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                     Maximum Streams (i)                     ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~

The fields in the MAX_STREAMS frame are as follows:

Maximum Streams:
: A count of the number of streams that can be opened.

> Loss or reordering can mean that a MAX_STREAMS frame can be received which

Reviewer 5: ```suggestion
Loss or reordering means that a MAX_STREAMS frame can be received which
```

Author: Fixed better, also used "endpoint" rather than client.

a change in the initial limits (see {{zerortt-parameters}}).

## MAX_STREAMS Frame {#frame-max-streams}

The MAX_STREAMS frame (type=0x1c and 0x1d) informs the peer of the number of
streams it is permitted to open.  A MAX_STREAMS frame with a type of 0x1c
applies to bidirectional streams; a MAX_STREAMS frame with a type of 0x1d

> applies to unidirectional streams.

Reviewer: Would it be clearer to frame this as a flag?

Author: I just tried this.  Describing using a flag means more words, and I think that it's a small regression in terms of clarity.

Reviewer 6: Yeah, I'm not thrilled with this either. How about we separate the frame types in the description? These are different frames, and I think it might be clearer to simply call them MAX_STREAMS_UNI and MAX_STREAMS_BIDI.

Reviewer 6: After having a look at #1900, I think its going to be hard to avoid this clunkiness. This PR is a net positive and combining these frames does improve readability, so I'll drop my suggestion.

Author: Thanks Jana.  There's no truly good way of dealing with this, but I think that we've at least hit a local minimum.

states a lower stream limit than the client has previously received.
MAX_STREAMS frames which do not increase the stream limit MUST be ignored.

> A peer MUST NOT open more streams than the limit it received permits.  For

Reviewer 2: Is that a sentence?

Reviewer: Yes, a correct albeit awkward one.  Perhaps "A peer MUST NOT open more streams than permitted by the current stream limit."?

a change in the initial limits (see {{zerortt-parameters}}).

## MAX_STREAMS Frame {#frame-max-streams}

The MAX_STREAMS frame (type=0x1c and 0x1d) informs the peer of the number of
streams it is permitted to open.  A MAX_STREAMS frame with a type of 0x1c
applies to bidirectional streams; a MAX_STREAMS frame with a type of 0x1d
applies to unidirectional streams.

The frame is as follows:

~~~
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                     Maximum Streams (i)                     ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~

The fields in the MAX_STREAMS frame are as follows:

Maximum Streams:

> : A count of the number of streams that can be opened.

Reviewer 5: ```suggestion
: A count of the total number of streams of a given type that can be opened over the course of the connection.
```

This is potentially ambiguous, because it could be read as "100 = 100 open streams at once."

Loss or reordering can cause a MAX_STREAMS frame to be received which states a
lower stream limit than an endpoint has previously received.  MAX_STREAMS frames
which do not increase the stream limit MUST be ignored.

> An endpoint MUST NOT open more streams than permitted by the current stream

Reviewer 5: ```suggestion
An endpoint MUST NOT open more total streams than permitted by the current stream
```

Stream Data Limit:

: A variable-length integer indicating the offset of the stream at which the
  blocking occurred.

## STREAMS_BLOCKED Frame {#frame-streams-blocked}

A sender SHOULD send a STREAMS_BLOCKED frame (type=0x1e or 0x1f) when it wishes
to open a stream, but is unable to due to the maximum stream limit set by its
peer (see {{frame-max-streams}}).  A STREAMS_BLOCKED frame of type 0x1e is used
to indicate reaching the bidirectional stream limit; a STREAMS_BLOCKED frame of
type 0x1f indicates reaching the unidirectional stream limit.

A STREAMS_BLOCKED frame does not open the stream, but informs the peer that a
new stream was needed and the stream limit prevented the creation of the stream.

The STREAMS_BLOCKED frame is as follows:

~~~
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                        Stream Limit (i)                     ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~

The STREAMS_BLOCKED frame contains a single field.

Stream Limit:

> : A variable-length integer indicating the stream limit at the time the frame

Reviewer 5: ```suggestion
: A variable-length integer indicating the limit on the total number of streams that may be opened in this connection.
```

: A variable-length integer indicating the stream which is flow control blocked.

Data Limit:

> : A variable-length integer indicating the offset of the stream at which the

Reviewer: Should we also use the term "stream-level limit" here?

  blocking occurred.

## STREAMS_BLOCKED Frame {#frame-streams-blocked}

A sender SHOULD send a STREAMS_BLOCKED frame (type=0x1e or 0x1f) when it wishes

> to open a stream, but is unable to due to the maximum stream ID limit set by its

Reviewer: s/ID//

peer (see {{frame-max-streams}}).  A STREAMS_BLOCKED frame of type 0x1e is used
to indicate reaching the bidirectional stream limit; a STREAMS_BLOCKED frame of

> type 0x1f indicates reaching the unidirectional stream limit.

Reviewer: Again, flag might be cleaner.

Reviewer 6: Similarly, UNI_STREAMS_BLOCKED and BIDI_STREAMS_BLOCKED.

Reviewer 6: Drop this suggestion too.
